test/mpi/rma/get_acc_flush.c

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+#include <stdlib.h>
+#include <assert.h>
+#include <unistd.h>
+#include <mpi.h>
+#include "mpitest.h"
+
+#define GET_BUF_SIZE 2147450880
+#define GET_SIGN_BUF_SIZE 16384 /* last MPI_Get data size */
+#define ACC_BUF_SIZE 32768
+#define SIGNATURE 0x77777777
+#define ITERATION 10
+
+/* This program tries to mimic the case where P0 issues rdma-get to remote
+ * P1, P2 issues AM-based accumulate to P0, P0 and P2 call MPI_Win_flush.
+ * MPI_Get will increase OFI internal rma counter of P0; during the call of
+ * MPIDI_OFI_PROGRESS, P0 may receive AM-based accumulate message from P2 and
+ * increase rma counter; however, after MPIDI_OFI_PROGRESS, issued_cntr (tcount) is
+ * not updated timely. This can cause incorrect exit from MPI_Win_flush with
+ * incomplete issued MPI_Get.
+ * In this test, we delay the handling of rdma-get on P1 by sleeping for 1 second,
+ * so it would allow accumulate issued by P2 to complete first and trigger issue_cntr
+ * update bug.
+ * The test assumes P0 is on node 0, and all others are on other nodes, so at least
+ * two nodes and three processes are required.
+ */
+
+int main(int argc, char *argv[])
+{
+    int rank, nprocs, target;
+    char *win_ptr, *acc_buf, *get_buf;
+    uint32_t *get_sign_buf;
+    MPI_Win win;
+    int get_cnt = 10;
+    int acc_cnt = 10;
+    int errs = 0;
+
+    MTest_Init(&argc, &argv);
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+    MPI_Comm_size(MPI_COMM_WORLD, &nprocs);
+
+    if (nprocs < 3) {
+        errs++;
+        printf("expected number of processes are at least 3\n");
+        goto exit;
+    }
+
+    if (rank == 0) {
+        target = 1;
+    } else if (rank == 2) {
+        target = 0;
+    }
+
+    if (rank < 3) {
+        MPI_Win_allocate(GET_BUF_SIZE, 1, MPI_INFO_NULL, MPI_COMM_WORLD, (void **) &win_ptr, &win);
+
+        MPI_Win_lock(MPI_LOCK_EXCLUSIVE, rank, 0, win);
+        memset(win_ptr, 0xaa, GET_BUF_SIZE);
+        MPI_Win_unlock(rank, win);
+        MPI_Barrier(MPI_COMM_WORLD);
+
+        acc_buf = (char *) malloc(ACC_BUF_SIZE);
+        memset(acc_buf, 0x01, ACC_BUF_SIZE);
+
+        get_buf = (char *) malloc(GET_BUF_SIZE);
+        memset(get_buf, 0, GET_BUF_SIZE);
+
+        get_sign_buf = (uint32_t *) malloc(GET_SIGN_BUF_SIZE);
+        memset(get_sign_buf, 0, GET_SIGN_BUF_SIZE);
+
+        MPI_Win_lock_all(MPI_MODE_NOCHECK, win);
+
+        for (int i = 0; i < ITERATION; ++i) {
+            if (rank == 0) {
+                *get_sign_buf = SIGNATURE;
+                for (int j = 0; j < get_cnt; ++j)
```

Comment:
Rather than checking whether `get_sign_buf ` has been overwritten, it is more accurate to check whether the value of `get_sign_buf ` has changed to the numbers set for win_ptr.